File: core/src/main/java/com/nao4j/subtrol/core/service/slice/MonthlySubscriptionSlicer.java
package com.nao4j.subtrol.core.service.slice;

import com.nao4j.subtrol.core.document.internal.ExactPeriod;
import com.nao4j.subtrol.core.document.internal.Quantity.QuantityType;
import com.nao4j.subtrol.core.document.internal.RightOpenPeriod;
import org.springframework.stereotype.Component;

import java.time.LocalDateTime;
import java.time.YearMonth;
import java.util.ArrayList;
import java.util.Collection;

import static com.nao4j.subtrol.core.document.internal.Quantity.QuantityType.MONTHS;
import static java.time.Period.between;
import static java.util.Collections.emptyList;

@Component
public class MonthlySubscriptionSlicer implements SubscriptionSlicer {

    @Override
    public Collection<ExactPeriod> slice(
            final RightOpenPeriod subscriptionPeriod,
            final ExactPeriod calculationPeriod,
            final long stepSize
    ) {
        if (isNotIntersect(subscriptionPeriod, calculationPeriod)) {
            return emptyList();
        }

        final var slices = new ArrayList<ExactPeriod>();
        var start = findRealCalculationStart(subscriptionPeriod, stepSize, calculationPeriod);
        while (isBeforeOrEquals(start, calculationPeriod.getEnd())
                && isBeforeOrEquals(start, subscriptionPeriod.getEnd())) {
            final var next = withDay(start.plusMonths(stepSize), subscriptionPeriod.getStart().getDayOfMonth());
            slices.add(new ExactPeriod(start, next));
            start = next;
        }
        return slices;
    }

    @Override
    public QuantityType dimension() {
        return MONTHS;
    }

    private boolean isNotIntersect(final RightOpenPeriod subscriptionPeriod, final ExactPeriod calculatePeriod) {
        return (subscriptionPeriod.getEnd() != null)
                && calculatePeriod.getStart().isAfter(subscriptionPeriod.getEnd())
                || calculatePeriod.getEnd().isBefore(subscriptionPeriod.getStart());
    }

    private boolean isBeforeOrEquals(final LocalDateTime value, final LocalDateTime reference) {
        return (reference == null) || value.isBefore(reference) || value.equals(reference);
    }

    private LocalDateTime findRealCalculationStart(
            final RightOpenPeriod subscriptionPeriod,
            final long stepSize,
            final ExactPeriod calculatePeriod
    ) {
        final var subscriptionStartDay = subscriptionPeriod.getStart().getDayOfMonth();

        final LocalDateTime start;
        if (subscriptionPeriod.getStart().isAfter(calculatePeriod.getStart())) {
            start = subscriptionPeriod.getStart();
        } else if (subscriptionStartDay < calculatePeriod.getStart().getDayOfMonth()) {
            start = withDay(calculatePeriod.getStart().plusMonths(stepSize), subscriptionStartDay);
        } else if (subscriptionStartDay >= calculatePeriod.getStart().getDayOfMonth()) {
            start = withDay(calculatePeriod.getStart(), subscriptionStartDay);
        } else {
            throw new IllegalArgumentException();   // todo: message
        }

        final var monthsToStart = between(subscriptionPeriod.getStart().toLocalDate(), start.toLocalDate()).getMonths();
        return start.plusMonths(monthsToStart % stepSize);
    }

    private LocalDateTime withDay(final LocalDateTime dateTime, final int day) {
        final var daysInMonth = YearMonth.from(dateTime).lengthOfMonth();
        if (day < daysInMonth) {
            return dateTime.withDayOfMonth(day);
        } else {
            return dateTime.withDayOfMonth(daysInMonth);
        }
    }

}

File: src/main/kotlin/com/nao4j/subtrol/service/slice/MonthlySubscriptionSlicer.kt
package com.nao4j.subtrol.service.slice

import com.nao4j.subtrol.model.ExactPeriod
import com.nao4j.subtrol.model.RightOpenPeriod
import java.time.LocalDateTime
import java.time.YearMonth

class MonthlySubscriptionSlicer : SubscriptionSlicer {

    override fun slice(
        subscriptionPeriod: RightOpenPeriod,
        calculationPeriod: ExactPeriod,
        stepSize: Long
    ): Collection<ExactPeriod> {
        if (isNotIntersect(subscriptionPeriod, calculationPeriod)) {
            return emptyList()
        }

        val slices = mutableListOf<ExactPeriod>()
        var start = findRealCalculationStart(subscriptionPeriod, stepSize, calculationPeriod)
        while (isBeforeOrEquals(start, calculationPeriod.end) && isBeforeOrEquals(start, subscriptionPeriod.end)) {
            val next = withDay(start.plusMonths(stepSize), subscriptionPeriod.start.dayOfMonth)
            slices.add(ExactPeriod(start, next))
            start = next
        }
        return slices
    }

    private fun isNotIntersect(subscriptionPeriod: RightOpenPeriod, calculatePeriod: ExactPeriod) =
        subscriptionPeriod.end != null && calculatePeriod.start.isAfter(subscriptionPeriod.end)
                || calculatePeriod.end.isBefore(subscriptionPeriod.start)

    private fun isBeforeOrEquals(value: LocalDateTime, reference: LocalDateTime?) =
        reference == null || value.isBefore(reference) || value == reference

    private fun findRealCalculationStart(
        subscriptionPeriod: RightOpenPeriod,
        stepSize: Long,
        calculatePeriod: ExactPeriod
    ): LocalDateTime {
        val subscriptionStartDay = subscriptionPeriod.start.dayOfMonth

        val start = when {
            (subscriptionPeriod.start.isAfter(calculatePeriod.start)) ->
                subscriptionPeriod.start
            (subscriptionStartDay < calculatePeriod.start.dayOfMonth) ->
                withDay(calculatePeriod.start.plusMonths(stepSize), subscriptionStartDay)
            (subscriptionStartDay >= calculatePeriod.start.dayOfMonth) ->
                withDay(calculatePeriod.start, subscriptionStartDay)
            else -> throw RuntimeException()
        }

        val monthsToStart = java.time.Period.between(subscriptionPeriod.start.toLocalDate(), start.toLocalDate()).months
        return start.plusMonths(monthsToStart % stepSize)
    }

    private fun withDay(dateTime: LocalDateTime, day: Int): LocalDateTime {
        val daysInMonth = YearMonth.from(dateTime).lengthOfMonth()
        return if (day < daysInMonth) {
            dateTime.withDayOfMonth(day)
        } else {
            dateTime.withDayOfMonth(daysInMonth)
        }
    }

}
